mirascope/tools/base.py

```diff
@@ -0,0 +1,89 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import ClassVar, Generic, TypeVar, cast
+
+from pydantic import BaseModel, ConfigDict, create_model
+
+from mirascope.core import BaseTool
+
+
+class _ToolConfig(BaseModel, ABC):
+    """Base configuration for tools"""
+
+    @classmethod
+    @abstractmethod
+    def from_env(cls) -> _ToolConfig:
+        """Load configuration from environment variables"""
+
+    @classmethod
+    def override_defaults(cls, config: _ToolConfig) -> _ToolConfig:
+        """Create config with overridden default values"""
+        return cls.model_validate(config)
+
+
+ConfigT = TypeVar("ConfigT", bound=_ToolConfig)
+
+_ToolSchemaT = TypeVar("_ToolSchemaT")
+
+
+class ConfigurableTool(BaseTool[_ToolSchemaT], Generic[ConfigT, _ToolSchemaT], ABC):
```

Comment:
why make a whole new class? Why not just extend `BaseTool` to have this functionality?

Reply:
From my understanding, I'm already extending (inheriting from) the existing `BaseTool`. While I'm indeed introducing new Generic variables due to typing constraints, isn't "extend" referring to class inheritance in this context?

Reply:
sorry, to phrase my comment better, I'm wondering why we aren't adding this functionality directly to `BaseTool`? I don't necessarily see the value in having two separate classes that we have to maintain?

Reply:
actually, scratch this. I think you're right to have the separate `ConfigurableTool` class.

We can also push this into `BaseTool` later if we want for whatever reason, but likely better to keep them separate for now